Stop using EnumTraits in OptionSet

achristensen07 opened the pull request:
#### 57afbafbe26097d0816e24c85b2a9c096bcab60d
<pre>
Stop using EnumTraits in OptionSet
<a href="https://bugs.webkit.org/show_bug.cgi?id=264903">https://bugs.webkit.org/show_bug.cgi?id=264903</a>

Reviewed by Timothy Hatcher.

It was used for only a few things:

1. It was used for isValidOptionSet, which is called when deserializing bytes from IPC.
   These uses are on their way to being entirely generated instead of using EnumTraits,
   and EnumTraits continues to be used for the types that are deserialized from IPC but
   have not yet been generated.
2. It was used in OptionSet::all to generate an OptionSet that contained only the valid
   bits.  The 3 uses of this function have been replaced by manually-written constexpr
   functions that are right next to the enum class definition, making it difficult to
   make errors of omission.  These 3 uses and any similar future uses are not worth
   keeping EnumTraits infrastructure around.
3. It was used to trim unused bits in OptionSet::fromRaw.  A manual inspection of all
   the functions named &quot;decodeForPersistence&quot; indicated that there are 0 places where
   OptionSet::fromRaw was used to &quot;parse&quot; data from disk.  Similarly, searching for
   &quot;&gt;::fromRaw(&quot; shows us the 36 places this is used, and most of them are just
   converting bit fields to OptionSets, but none of them are used to get data from an
   untrusted source and use operations like OptionSet::operator==.  Other operations
   like contains, add, remove, etc. are unaffected by unrecognized bits.
4. It was used for debug assertions in constructors that take enum values directly.  I
   replaced these assertions by stricter assertions using hasOneBitSet instead.

Generated serialization is generating isValidOptionSet functions directly, which keeps
our code from decoding unexpected bits from IPC to prevent untrusted, possibly
compromised web content processes from being able to do unexpected things.

* Source/WTF/wtf/OptionSet.h:
(WTF::OptionSet::fromRaw):
(WTF::OptionSet::OptionSet):
(WTF::isValidOptionSet):
(WTF::isValidOptionSetEnum): Deleted.
(WTF::maskRawValue): Deleted.
(WTF::OptionSet::all): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(toImpl):
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
(WebKit::WebExtensionTab::allChangedProperties):
* Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h:
(WebKit::allWebExtensionWindowTypeFilters):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::WebExtensionAPIWindows::parseWindowTypesFilter):
(WebKit::WebExtensionContextProxy::dispatchWindowsEvent):

Canonical link: <a href="https://commits.webkit.org/270838@main">https://commits.webkit.org/270838@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/a7ca513422153a803f745c3de4b96c9609278fdf

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [✅ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/26568 "Passed style check") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/48/builds/5184 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/27816 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/28780 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/24298 "Built successfully") 
| | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/49/builds/6993 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/51/builds/2591 "Built successfully") | [  ~~🧪 wpe-wk2~~](https://ews-build.webkit.org/#/builders/34/builds/24259 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/26829 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/47/builds/4017 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/22829 "Passed tests") | [✅ 🧪 api-wpe](https://ews-build.webkit.org/#/builders/41/builds/3541 "Passed tests") | 
| | [  ~~🧪 ios-wk2-wpt~~](https://ews-build.webkit.org/#/builders/42/builds/3579 "The change is no longer eligible for processing. Pull Request was already closed when EWS attempted to process it.") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/23815 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/29262 "Built successfully") | 
| [✅ 🛠 🧪 jsc](https://ews-build.webkit.org/#/builders/20/builds/23152 "Built successfully and passed tests") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/24231 "Passed tests") | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/24214 "Passed tests") | [  ~~🧪 gtk-wk2~~](https://ews-build.webkit.org/#/builders/1/builds/29851 "The change is no longer eligible for processing. Commit was outdated when EWS attempted to process it.") | 
| [✅ 🛠 🧪 jsc-arm64](https://ews-build.webkit.org/#/builders/12/builds/25773 "Built successfully and passed tests") | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/44/builds/3612 "Built successfully") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/50/builds/1815 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/27748 "Passed tests") | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/46/builds/5048 "Built successfully") | | [✅ 🛠 jsc-armv7](https://ews-build.webkit.org/#/builders/35/builds/33228 "Built successfully") | 
| [✅ 🛠 🧪 merge](https://ews-build.webkit.org/#/builders/19/builds/6372 "Built successfully and passed tests") | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/43/builds/4086 "Built successfully") | | [✅ 🧪 jsc-armv7-tests](https://ews-build.webkit.org/#/builders/25/builds/7196 "Passed tests") | 
| | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/45/builds/3940 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

xeenon commented on Source/WTF/wtf/OptionSet.h:
```diff
@@ -242,10 +184,22 @@ template<typename E> class OptionSet {
 WARN_UNUSED_RETURN constexpr bool isValidOptionSet(OptionSet<E> optionSet)
 {
-    auto allValidBitsValue = OptionSetValueChecker<std::make_unsigned_t<std::underlying_type_t<E>>, typename EnumTraits<E>::values>::allValidBits();
+    auto allValidBitsValue = IsValidOptionSetHelper::OptionSetValueChecker<std::make_unsigned_t<std::underlying_type_t<E>>, typename EnumTraits<E>::values>::allValidBits();
```
Does this EnumTraits use need removed?

achristensen07 replied:
It does, but not until all deserialized OptionSet enums are migrated to generated serialization.  I added a FIXME comment indicating such.

xeenon commented on Source/WTF/wtf/OptionSet.h:
```diff
@@ -130,26 +77,21 @@ template<typename E> class OptionSet {
         : m_storage(static_cast<StorageType>(e))
     {
-        ASSERT(!m_storage || isValidOptionSetEnum(e));
+        ASSERT(!m_storage || hasOneBitSet(static_cast<typename OptionSet<E>::StorageType>(e)));
```
Wouldn't `typename StorageType` work?

cdumez replied:
Heck, I think `static_cast<StorageType>(e)` would work.

xeenon commented on Source/WTF/wtf/OptionSet.h:
```diff
@@ -130,26 +77,21 @@ template<typename E> class OptionSet {
     {
         for (auto& option : initializerList) {
-            ASSERT(isValidOptionSetEnum(option));
+            ASSERT(hasOneBitSet(static_cast<typename OptionSet<E>::StorageType>(option)));
```
Ditto for `typename StorageType`?

cdumez replied:
ditto

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
```diff
@@ -77,6 +77,20 @@ class WebExtensionTab : public RefCounted<WebExtensionTab>, public CanMakeWeakPt
         URL        = 1 << 8,
         ZoomFactor = 1 << 9,
     };
+    static constexpr OptionSet<ChangedProperties> allChangedProperties()
```
Nit: Add a new line above.

xeenon commented on Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h:
```diff
@@ -49,6 +49,13 @@ enum class WebExtensionWindowTypeFilter : uint8_t {
     Normal = 1 << 0,
     Popup  = 1 << 1,
 };
+static constexpr OptionSet<WebExtensionWindowTypeFilter> allWebExtensionWindowTypeFilters()
```
Nit: New line above.

achristensen07 committed:
Stop using EnumTraits in OptionSet
https://bugs.webkit.org/show_bug.cgi?id=264903

Reviewed by Timothy Hatcher.

It was used for only a few things:

1. It was used for isValidOptionSet, which is called when deserializing bytes from IPC.
   These uses are on their way to being entirely generated instead of using EnumTraits,
   and EnumTraits continues to be used for the types that are deserialized from IPC but
   have not yet been generated.
2. It was used in OptionSet::all to generate an OptionSet that contained only the valid
   bits.  The 3 uses of this function have been replaced by manually-written constexpr
   functions that are right next to the enum class definition, making it difficult to
   make errors of omission.  These 3 uses and any similar future uses are not worth
   keeping EnumTraits infrastructure around.
3. It was used to trim unused bits in OptionSet::fromRaw.  A manual inspection of all
   the functions named "decodeForPersistence" indicated that there are 0 places where
   OptionSet::fromRaw was used to "parse" data from disk.  Similarly, searching for
   ">::fromRaw(" shows us the 36 places this is used, and most of them are just
   converting bit fields to OptionSets, but none of them are used to get data from an
   untrusted source and use operations like OptionSet::operator==.  Other operations
   like contains, add, remove, etc. are unaffected by unrecognized bits.
4. It was used for debug assertions in constructors that take enum values directly.  I
   replaced these assertions by stricter assertions using hasOneBitSet instead.

Generated serialization is generating isValidOptionSet functions directly, which keeps
our code from decoding unexpected bits from IPC to prevent untrusted, possibly
compromised web content processes from being able to do unexpected things.

* Source/WTF/wtf/OptionSet.h:
(WTF::OptionSet::fromRaw):
(WTF::OptionSet::OptionSet):
(WTF::isValidOptionSet):
(WTF::isValidOptionSetEnum): Deleted.
(WTF::maskRawValue): Deleted.
(WTF::OptionSet::all): Deleted.
* Source/WebKit/UIProcess/API/Cocoa/_WKWebExtensionContext.mm:
(toImpl):
* Source/WebKit/UIProcess/Extensions/WebExtensionTab.h:
(WebKit::WebExtensionTab::allChangedProperties):
* Source/WebKit/UIProcess/Extensions/WebExtensionWindow.h:
(WebKit::allWebExtensionWindowTypeFilters):
* Source/WebKit/WebProcess/Extensions/API/Cocoa/WebExtensionAPIWindowsCocoa.mm:
(WebKit::WebExtensionAPIWindows::parseWindowTypesFilter):
(WebKit::WebExtensionContextProxy::dispatchWindowsEvent):

Canonical link: https://commits.webkit.org/270838@main
